src/main/java/com/redhat/foreman/DisposableImpl.java

```diff
@@ -0,0 +1,40 @@
+package com.redhat.foreman;
+
+import org.apache.log4j.Logger;
+import org.jenkinsci.plugins.resourcedisposer.Disposable;
+
+import javax.annotation.Nonnull;
+
+/**
+ * Created by shebert on 27/09/16.
+ */
+public class DisposableImpl implements Disposable {
+
+    private final String cloudName;
+    private final String name;
+    private static final Logger LOGGER = Logger.getLogger(DisposableImpl.class);
+
+    public DisposableImpl(String cloudName, String name) {
+        this.cloudName = cloudName;
+        this.name = name;
+    }
+
+    @Nonnull
+    @Override
+    public State dispose() throws Exception {
+        ForemanSharedNodeCloud cloud = ForemanSharedNodeCloud.getByName(cloudName);
+        if (cloud != null) {
+            cloud.getForemanAPI().release(name);
```

Comment:
Note the plugin expects this will not throw in case the machine get unregistered/deleted externally. So your disposable will not hang in queue failing with a variant of `NotFound` exception forever.